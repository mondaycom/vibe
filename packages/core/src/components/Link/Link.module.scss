@import "../../styles/typography";
@import "~monday-ui-style/dist/mixins";

.link {
  display: flex;
  align-items: center;
  color: var(--link-color);
  @include vibe-text(text2, normal);
  @include smoothing-text;
  text-decoration: none;
  writing-mode: horizontal-tb;
  @include focus-style();

  &.colorOnPrimary {
    color: var(--text-color-on-primary);
  }

  &.colorOnInverted {
    color: var(--text-color-on-inverted);
  }
}

.link:hover .text {
  text-decoration: underline;
  color: inherit;
}

.link.inlineText {
  display: inline-flex;
  align-items: unset;
}

.link.inheritFontSize {
  font-size: inherit;
  line-height: inherit;
}

.iconStart {
  line-height: 24px;
  margin-right: var(--spacing-small);
}

.iconEnd {
  line-height: 24px;
  margin-left: var(--spacing-small);
}

@supports (margin-inline-start: initial) {
  .iconStart {
    line-height: 24px;
    margin-right: 0;
    margin-inline-end: var(--spacing-xs);
  }
  .iconEnd {
    line-height: 24px;
    margin-left: 0;
    margin-inline-start: var(--spacing-xs);
  }
}
