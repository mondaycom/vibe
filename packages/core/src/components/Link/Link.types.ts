export type LinkTarget = "_blank" | "_self" | "_parent" | "_top";

export type LinkIconPosition = "start" | "end";

export type LinkColor = "primary" | "onPrimary" | "onInverted";
