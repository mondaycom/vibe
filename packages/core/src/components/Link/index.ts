export { default as Link, type LinkProps } from "./Link";

export * from "./Link.types";
