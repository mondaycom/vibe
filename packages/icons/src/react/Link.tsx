/* eslint-disable */
/* tslint:disable */
import * as React from 'react';
export interface LinkProps extends React.SVGAttributes<SVGElement> {
size?: string | number;
}
const Link: React.FC<LinkProps> = ({size, ...props}) => (
  <svg viewBox="0 0 20 20" fill="currentColor" width={ size || "20" } height={ size || "20" } {...props}>
    <path d="M7.3125 5.3125C7.72671 5.3125 8.0625 5.64829 8.0625 6.0625C8.0625 6.47671 7.72671 6.8125 7.3125 6.8125H6.8125C5.00032 6.8125 3.53125 8.28157 3.53125 10.0938C3.53125 11.9059 5.00032 13.375 6.8125 13.375H7.3125C7.72671 13.375 8.0625 13.7108 8.0625 14.125C8.0625 14.5392 7.72671 14.875 7.3125 14.875H6.8125C4.17189 14.875 2.03125 12.7344 2.03125 10.0938C2.03125 7.45314 4.17189 5.3125 6.8125 5.3125H7.3125ZM11.9688 6.0625C11.9688 5.64829 12.3045 5.3125 12.7188 5.3125H13.1875C15.8281 5.3125 17.9688 7.45314 17.9688 10.0938C17.9688 12.7344 15.8281 14.875 13.1875 14.875H12.7188C12.3045 14.875 11.9688 14.5392 11.9688 14.125C11.9688 13.7108 12.3045 13.375 12.7188 13.375H13.1875C14.9997 13.375 16.4688 11.9059 16.4688 10.0938C16.4688 8.28157 14.9997 6.8125 13.1875 6.8125H12.7188C12.3045 6.8125 11.9688 6.47671 11.9688 6.0625ZM5.85834 10.079C5.85834 9.66477 6.19412 9.32898 6.60834 9.32898H13.3604C13.7746 9.32898 14.1104 9.66477 14.1104 10.079C14.1104 10.4932 13.7746 10.829 13.3604 10.829H6.60834C6.19412 10.829 5.85834 10.4932 5.85834 10.079Z"
      fill="currentColor" fillRule="evenodd" clipRule="evenodd" />
  </svg>
);
Link.displayName = 'Link';
export default Link;
/* tslint:enable */
/* eslint-enable */
