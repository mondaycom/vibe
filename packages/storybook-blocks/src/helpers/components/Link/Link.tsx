import cx from 'classnames';
import React, { forwardRef, useCallback } from 'react';
import { NOOP } from '../../utils/function-utils';
import { LinkTarget } from './LinkConsts';
import { VibeComponentProps, VibeComponent } from '../../types';
import { withStaticProps } from '../../../types';
import styles from './Link.module.scss';

interface LinkProps extends VibeComponentProps {
  /**
   * Class name for overriding link text styles
   */
  textClassName?: string;
  /** Specifies the location (URL) of the external resource */
  href?: string;
  /** The link text */
  text?: React.ReactNode | React.ReactNode[] | string;
  /** Defines the relationship between a linked resource and the current document */
  rel?: string;
  /** onClick function - MouseEvent */
  onClick?: (event: React.MouseEvent) => void;
  /** Specifies where to open the linked document */
  target?: LinkTarget;
  /** Aria label description */
  ariaLabelDescription?: string;
  /** element id to describe the counter accordingly */
  ariaLabeledBy?: string;
  /** disable navigation */
  disableNavigation?: boolean;
  /** inherit text size */
  inheritFontSize?: boolean;
  /** if the link is in part of other text content */
  inlineText?: boolean;
}

const Link: VibeComponent<LinkProps, HTMLAnchorElement> & {
  targets?: typeof LinkTarget;
  target?: typeof LinkTarget;
} = forwardRef(
  (
    {
      className,
      textClassName,
      href = '',
      text = '',
      rel = 'noreferrer',
      onClick = NOOP,
      target = Link.targets?.NEW_WINDOW,
      ariaLabelDescription = '',
      id = '',
      ariaLabeledBy = '',
      disableNavigation = false,
      inheritFontSize = false,
      inlineText = false,
    },
    ref: React.ForwardedRef<HTMLAnchorElement>,
  ) => {
    const onClickWrapper = useCallback(
      (e: React.MouseEvent<HTMLElement>) => {
        if (disableNavigation) {
          e.preventDefault();
        }
        onClick && onClick(e);
      },
      [disableNavigation, onClick],
    );

    return (
      <a
        id={id}
        href={href}
        rel={rel}
        ref={ref}
        onClick={onClickWrapper}
        target={target}
        className={cx(styles.link, className, {
          [styles.inheritFontSize]: inheritFontSize,
          [styles.inlineText]: inlineText,
        })}
        aria-label={ariaLabelDescription}
        aria-labelledby={ariaLabeledBy}
      >
        <span className={cx(styles.text, textClassName)}>{text}</span>
      </a>
    );
  },
);

export default withStaticProps(Link, {
  target: LinkTarget,
  targets: LinkTarget,
});
