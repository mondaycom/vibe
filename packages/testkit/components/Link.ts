import { Locator, Page, test } from "@playwright/test";
import { BaseElement } from "./BaseElement";

/**
 * Class representing a Link element.
 * Extends the BaseElement class.
 */
export class Link extends BaseElement {
  /**
   * Create a Link element.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the Link element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
  }

  /**
   * Get the href attribute of the link.
   * @returns {Promise<string>} The href attribute of the link.
   */
  async getLinkHref(): Promise<string> {
    return await test.step(`Get link href of ${this.getElementReportName()}`, async () => {
      return await this.getAttributeValue("href");
    });
  }
}
