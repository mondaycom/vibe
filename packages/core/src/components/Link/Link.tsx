import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import cx from "classnames";
import React, { forwardRef, useCallback } from "react";
import { NOOP } from "../../utils/function-utils";
import Icon from "../Icon/Icon";
import { IconPosition as IconPositionEnum, LinkTarget as LinkTargetEnum } from "./LinkConsts";
import { LinkColor, LinkIconPosition, LinkTarget } from "./Link.types";
import { SubIcon, VibeComponentProps, withStaticProps } from "../../types";
import styles from "./Link.module.scss";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import { camelCase } from "lodash-es";

export interface LinkProps extends VibeComponentProps {
  /**
   * Class name applied to the link text.
   */
  textClassName?: string;
  /**
   * The URL the link points to.
   */
  href?: string;
  /**
   * The text content of the link.
   */
  text?: string;
  /**
   * Specifies the relationship between the current document and the linked resource.
   */
  rel?: string;
  /**
   * Callback fired when the link is clicked.
   */
  onClick?: (event: React.MouseEvent) => void;
  /**
   * Specifies where to open the linked document.
   */
  target?: LinkTarget;
  /**
   * The ARIA label description for accessibility.
   */
  ariaLabelDescription?: string;
  /**
   * The ID of the element that describes this link.
   */
  ariaDescribedby?: string;
  /**
   * The ID of the element labeling this link.
   */
  ariaLabeledBy?: string;
  /**
   * Icon displayed next to the link text.
   */
  icon?: SubIcon;
  /**
   * The position of the icon relative to the text.
   */
  iconPosition?: LinkIconPosition;
  /**
   * If true, disables navigation when the link is clicked.
   */
  disableNavigation?: boolean;
  /**
   * If true, the link inherits the surrounding text's font size.
   */
  inheritFontSize?: boolean;
  /**
   * If true, the link is styled to fit within inline text content.
   */
  inlineText?: boolean;
  /** The link's color style */
  color?: LinkColor;
}

const Link = forwardRef(
  (
    {
      className,
      textClassName,
      href = "",
      text = "",
      rel = "noreferrer",
      onClick = NOOP,
      target = "_blank",
      ariaLabelDescription = "",
      color = "primary",
      ariaDescribedby = "",
      icon = "",
      iconPosition = "start",
      id = "",
      ariaLabeledBy = "",
      disableNavigation = false,
      inheritFontSize = false,
      inlineText = false,
      "data-testid": dataTestId
    }: LinkProps,
    ref: React.ForwardedRef<HTMLAnchorElement>
  ) => {
    const isStart = iconPosition === "start";

    const onClickWrapper = useCallback(
      (e: React.MouseEvent<HTMLElement>) => {
        if (disableNavigation) {
          e.preventDefault();
        }
        onClick && onClick(e);
      },
      [disableNavigation, onClick]
    );

    return (
      <a
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.LINK, id)}
        id={id}
        href={href}
        rel={rel}
        ref={ref}
        onClick={onClickWrapper}
        target={target}
        className={cx(styles.link, className, getStyle(styles, camelCase("color-" + color)), {
          [styles.inheritFontSize]: inheritFontSize,
          [styles.inlineText]: inlineText
        })}
        aria-label={ariaLabelDescription}
        aria-describedby={ariaDescribedby}
        aria-labelledby={ariaLabeledBy}
      >
        {getIcon(isStart, icon, cx(styles.iconStart))}
        <span className={cx(styles.text, textClassName)}>{text}</span>
        {getIcon(!isStart, icon, cx(styles.iconEnd))}
      </a>
    );
  }
);

function getIcon(shouldShow: boolean, icon: string | React.FunctionComponent | null, className: string) {
  if (!shouldShow) return;
  return <Icon className={className} icon={icon} iconType="font" />;
}

interface LinkStaticProps {
  targets: typeof LinkTargetEnum;
  iconPositions: typeof IconPositionEnum;
  position: typeof IconPositionEnum;
}

export default withStaticProps<LinkProps, LinkStaticProps>(Link, {
  position: IconPositionEnum,
  iconPositions: IconPositionEnum,
  targets: LinkTargetEnum
});
