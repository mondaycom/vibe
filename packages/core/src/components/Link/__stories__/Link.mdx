import { Meta } from "@storybook/blocks";
import Link from "../Link.tsx";
import { ExternalPage, Gallery } from "@vibe/icons";
import Example from "./Example";
import { BREADCRUBMS, BUTTON, BADGE } from "../../../storybook/components/related-components/component-description-map";
import { TipButton } from "./Link.stories.helpers";
import * as LinkStories from "./Link.stories";
import tipseen1 from "./tipseen1.svg";
import tipseen2 from "./tipseen2.svg";
import tipseen3 from "./tipseen3.svg";
import tipseen4 from "./tipseen4.svg";

<Meta of={LinkStories} />

# Link

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Use cases and examples](#use-cases-and-examples)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Link is an actionable text component with connection to another web pages.

<Canvas of={LinkStories.Overview} />

### Import

```js
import { Link } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Use links only as part of the text or scentence.",
    "Use links in order to redirect people outside of the system.",
    "Links should be short with a clear action name. read more, go to, etc..."
  ]}
/>

<TipButton />

**Link's color on Primary**

- The link on Primary surface should be `text-color-on-primary` with underline.

- For hover state the cursor should transit to pointer to represent clickable area.

<Example>
  <img src={tipseen1} alt="Example of link on primary surface" />
  <img src={tipseen2} alt="Example of link on primary surface on hover" />
</Example>

**Link's color on inverted**

- The link on inverted surface should be `text-color-on-inverted` with underline.

- For hover state the cursor should transit to pointer to represent clickable area.

<Example>
  <img src={tipseen3} alt="Example of link on inverted surface" />
  <img src={tipseen4} alt="Example of link on inverted surface on hover" />
</Example>

## Variants

### States

<Canvas of={LinkStories.States} />

### Right to left

<Canvas of={LinkStories.RightToLeft} />

### With icons

<Canvas of={LinkStories.WithIcons} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: <Link text="Read more" href="https://www.monday.com" />,
        description: "Keep the copy short and to the point."
      },
      negative: {
        component: <Link text="Go to this url and read all about it" href="https://www.monday.com" />,
        description: "Don’t write long messages that are clickable."
      }
    },
    {
      positive: {
        component: <Link text="Read more" iconPosition="start" href="https://www.monday.com" icon={ExternalPage} />,
        description: "Use icon from the left or right side of the link"
      },
      negative: {
        component: [
          <Link icon={Gallery} iconPosition="end" />,
          <Link icon={ExternalPage} text="Read more" iconPosition="end" />
        ],
        description: "Don’t use two icons for one link."
      }
    },
    {
      positive: {
        component: <Link text="Read more" href="https://www.monday.com" />,
        description: "Use only one link in context in text"
      },
      negative: {
        component: [
          <Link text="Read this" href="https://www.monday.com" />,
          <span class="monday-storybook-link_text">or</span>,
          <Link text="this" href="https://www.monday.com" />,
          <Link text="Nice article" href="https://www.monday.com" />
        ],
        description: "Don’t use multiple links near each other"
      }
    }
  ]}
/>

## Use cases and examples

### Reference link

Use this menu to allow a user to either select one or more items from the list.

<Canvas of={LinkStories.ReferenceLink} />

### Shortening texts

Use read more to shorten long paragraphs of text

<Canvas of={LinkStories.ShorteningTexts} />

## Related components

<RelatedComponents componentsNames={[BUTTON, BREADCRUBMS, BADGE]} />
