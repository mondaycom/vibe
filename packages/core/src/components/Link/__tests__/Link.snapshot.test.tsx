import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import Link from "../Link";

describe("Link renders correctly", () => {
  it("with icon", () => {
    const tree = renderer.create(<Link icon="fa fa-star" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with right icon", () => {
    const tree = renderer.create(<Link icon="fa fa-star" iconPosition="end" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with id", () => {
    const tree = renderer.create(<Link id="TestId" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with className", () => {
    const tree = renderer.create(<Link className="testClassName" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with text", () => {
    const tree = renderer.create(<Link text="Link" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with ariaLabeledBy", () => {
    const tree = renderer.create(<Link ariaLabeledBy="aria label link" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with ariaDescribedby", () => {
    const tree = renderer.create(<Link ariaDescribedby="aria described by" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with ariaLabelDescription", () => {
    const tree = renderer.create(<Link ariaLabelDescription="arialabel link" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
