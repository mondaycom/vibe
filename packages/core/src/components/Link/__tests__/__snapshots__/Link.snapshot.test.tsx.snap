// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly with ariaDescribedby 1`] = `
<a
  aria-describedby="aria described by"
  aria-label=""
  aria-labelledby=""
  className="link colorPrimary"
  data-testid="link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with ariaLabelDescription 1`] = `
<a
  aria-describedby=""
  aria-label="arialabel link"
  aria-labelledby=""
  className="link colorPrimary"
  data-testid="link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with ariaLabeledBy 1`] = `
<a
  aria-describedby=""
  aria-label=""
  aria-labelledby="aria label link"
  className="link colorPrimary"
  data-testid="link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with className 1`] = `
<a
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className="link testClassName colorPrimary"
  data-testid="link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with icon 1`] = `
<a
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className="link colorPrimary"
  data-testid="link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    aria-hidden={true}
    className="icon iconStart fa fa fa-star"
    data-testid="icon"
    data-vibe="Icon"
    onClick={[Function]}
    role="img"
  />
  <span
    className="text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with id 1`] = `
<a
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className="link colorPrimary"
  data-testid="link_TestId"
  href=""
  id="TestId"
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with right icon 1`] = `
<a
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className="link colorPrimary"
  data-testid="link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="text"
  >
    
  </span>
  <span
    aria-hidden={true}
    className="icon iconEnd fa fa fa-star"
    data-testid="icon"
    data-vibe="Icon"
    onClick={[Function]}
    role="img"
  />
</a>
`;

exports[`Link renders correctly with text 1`] = `
<a
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className="link colorPrimary"
  data-testid="link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="text"
  >
    Link
  </span>
</a>
`;
